--- batman-adv-2018.3/net/batman-adv/fast_batman.c	2018-11-09 16:57:00.706463585 +0800
+++ batman-adv-2018.3/net/batman-adv/fast_batman.c	2018-11-09 17:21:24.006460494 +0800
@@ -77,6 +77,7 @@
 #endif
 static void	fastbat_del_local_entry(FAST_LOCAL *);
 static void	fastbat_show_local_entry(void);
+static void	fastbat_local_add_check_global_and_setup_dirty(u8 *);
 
 /* sysctl variable to turn on/off fastbat dynamically */
 extern int fastmesh_enable;
@@ -456,6 +457,7 @@
 	if(vlan)
 		local->vlan=vlan;
 #endif
+	fastbat_local_add_check_global_and_setup_dirty(local->dst);
 	list_add(local, &fastbat_local_head);
 
 RETN:
@@ -1495,7 +1497,7 @@
 		queue_delayed_work(batadv_event_workqueue, &fastbat_work, msecs_to_jiffies(FASTBAT_DIRTY_PURGE));
 }
 
-void	fastbat_check_disconnect_neighbor_and_setup_dirty
+void	fastbat_local_add_check_global_and_setup_dirty
 (u8 *addr)
 {
 	FAST_BAT	*bat;
@@ -1506,6 +1508,34 @@
 	for(; idx<FAST_BAT_LIST_LEN; idx++) {
 		hlist_for_each_entry(bat, &bats[idx], list) {
 			if(bat->dirty)
+				continue;
+#if	!defined(COLLECT_ORIG_OR_NEIGH)
+			if(!memcmp(addr, bat->dst, ETH_ALEN)) {
+#else
+			if(!memcmp(addr, bat->dst, ETH_ALEN-1)) {
+#endif
+				bat->dirty=1;
+				dirty=1;
+				FASTBAT_DBG_RAW_MSG("[F@STBAT] find out:(o)%pM\n (s)%pM\n (d)%pM\n (rs)%pM\n (rd)%pM\n", 
+					bat->orig, bat->src1, bat->dst1, bat->src, bat->dst);
+			}
+		}
+	}
+	if(dirty)
+		queue_delayed_work(batadv_event_workqueue, &fastbat_work, msecs_to_jiffies(FASTBAT_DIRTY_PURGE));	
+}
+
+void	fastbat_check_disconnect_neighbor_and_setup_dirty
+(u8 *addr)
+{
+	FAST_BAT	*bat;
+	u32			idx=0;
+	int			dirty=0;
+
+	FASTBAT_DBG_RAW_MSG("[F@STBAT] setup dirty from disconnect (o)%pM\n", addr);
+	for(; idx<FAST_BAT_LIST_LEN; idx++) {
+		hlist_for_each_entry(bat, &bats[idx], list) {
+			if(bat->dirty)
 				continue;
 			//if(!memcmp(addr, bat->dst1, ETH_ALEN) || !memcmp(addr, bat->src1, ETH_ALEN)) {
 #if	!defined(COLLECT_ORIG_OR_NEIGH)
