--- batman-adv-2018.3/net/batman-adv/send.c	2019-06-28 12:13:01.040868083 +0800
+++ batman-adv-2018.3/net/batman-adv/send.c	2019-06-28 11:41:26.464909310 +0800
@@ -55,6 +55,7 @@
 static void batadv_send_outstanding_bcast_packet(struct work_struct *work);
 
 #if	defined(CONFIG_BATMAN_ADV_FASTBAT)
+extern int fastmesh_enable;
 extern int fastbat_chk_and_add_list_entry(struct sk_buff *, struct net_device *, struct net_device *, u8 *);
 #endif
 
@@ -122,7 +123,8 @@
 	 * (which is > 0). This will not be treated as an error.
 	 */
 #if	defined(CONFIG_BATMAN_ADV_FASTBAT)
-	fastbat_chk_and_add_list_entry(skb, hard_iface->soft_iface, hard_iface->net_dev, dst_addr);
+	if(fastmesh_enable)
+		fastbat_chk_and_add_list_entry(skb, hard_iface->soft_iface, hard_iface->net_dev, dst_addr);
 #endif
 	ret = dev_queue_xmit(skb);
 	return net_xmit_eval(ret);
--- batman-adv-2018.3/net/batman-adv/fast_batman.c	2019-06-28 12:13:01.296868077 +0800
+++ batman-adv-2018.3/net/batman-adv/fast_batman.c	2019-06-28 11:47:16.296901698 +0800
@@ -29,7 +29,11 @@
 #define xxxFASTBAT_DBG_RAW_MSG(...)
 #define FASTBAT_DBG_MSG(...)
 #define xxxFASTBAT_DBG_MSG(...)
-#define FASTBAT_ERR_MSG(...)                    printk(KERN_ERR __VA_ARGS__)
+#define FASTBAT_ERR_MSG(...)					\
+do {											\
+	if(__ratelimit(&fastbat_err_state))			\
+		printk(KERN_ERR "[F@STBAT]" __VA_ARGS__);	\
+} while(0)
 #endif
 
 #define FASTBAT_WARN_ON(...)					\
@@ -727,13 +731,13 @@
 #endif
 		if(orig_node != NULL) {
 			if(memcmp(bat->src1, neigh, ETH_ALEN-1) == 0) {
-				FASTBAT_ERR_MSG("[F@STBAT] looping!!!\n");
+				FASTBAT_ERR_MSG(" looping!!!\n");
 				bat->dirty=1;
 			}
 			ether_addr_copy(bat->orig, orig_node->orig);
 #if	defined(COLLECT_ORIG_OR_NEIGH)
 		} else {
-			FASTBAT_ERR_MSG("[F@STBAT] no orig!!!\n");
+			FASTBAT_ERR_MSG(" no orig!!!\n");
 			bat->dirty=1;
 			ether_addr_copy(bat->orig, neigh);
 #endif
